File: patterns/_tests/BehavioralPatterns/Observer/csharp/unit.test.cs
// ============================================================
//  Unit Tests — Observer: Weather Station (C#)
//
//  What we test here:
//    · Observer registration/removal mechanics
//    · Notification broadcast to all registered observers
//    · Pull model — observers get current readings
//    · Reading history maintained correctly
//    · Observer independence — removal doesn't affect others
//
//  Run: dotnet test
// ============================================================

using System;
using System.Collections.Generic;
using Xunit;


class MockObserver : Observer
{
    public int UpdateCount { get; set; } = 0;
    public WeatherStation LastStation { get; set; }

    public void Update(WeatherStation station)
    {
        UpdateCount++;
        LastStation = station;
    }
}


public class UnitTest
{
    private WeatherStation _station;
    private MockObserver _observer1;
    private MockObserver _observer2;

    public UnitTest()
    {
        _station = new WeatherStation();
        _observer1 = new MockObserver();
        _observer2 = new MockObserver();
    }

    // ── Observer Registration Tests
    [Fact]
    public void TestRegistersASingleObserver()
    {
        _station.RegisterObserver(_observer1);
        Assert.Equal(0, _observer1.UpdateCount); // No update yet
    }

    [Fact]
    public void TestPreventsDuplicateRegistration()
    {
        _station.RegisterObserver(_observer1);
        _station.RegisterObserver(_observer1); // Try again
        _station.SetWeatherData(20, 60, 1010, 10);
        Assert.Equal(1, _observer1.UpdateCount); // Called only once
    }

    [Fact]
    public void TestRemovesAnObserverFromTheList()
    {
        _station.RegisterObserver(_observer1);
        _station.RegisterObserver(_observer2);
        _station.RemoveObserver(_observer1);
        _station.SetWeatherData(20, 60, 1010, 10);
        Assert.Equal(0, _observer1.UpdateCount); // Not notified
        Assert.Equal(1, _observer2.UpdateCount); // Still notified
    }

    // ── Notification Broadcast Tests
    [Fact]
    public void TestNotifiesAllRegisteredObservers()
    {
        var obs3 = new MockObserver();
        _station.RegisterObserver(_observer1);
        _station.RegisterObserver(_observer2);
        _station.RegisterObserver(obs3);

        _station.SetWeatherData(25, 70, 1015, 15);

        Assert.Equal(1, _observer1.UpdateCount);
        Assert.Equal(1, _observer2.UpdateCount);
        Assert.Equal(1, obs3.UpdateCount);
    }

    [Fact]
    public void TestNotifiesRepeatedlyOnMultipleUpdates()
    {
        _station.RegisterObserver(_observer1);

        _station.SetWeatherData(20, 60, 1010, 10);
        Assert.Equal(1, _observer1.UpdateCount);

        _station.SetWeatherData(25, 70, 1015, 15);
        Assert.Equal(2, _observer1.UpdateCount);

        _station.SetWeatherData(30, 80, 1020, 20);
        Assert.Equal(3, _observer1.UpdateCount);
    }

    [Fact]
    public void TestPassesTheStationItselfToUpdate()
    {
        _station.RegisterObserver(_observer1);
        _station.SetWeatherData(25, 70, 1015, 15);

        Assert.Same(_observer1.LastStation, _station);
    }

    // ── Pull Model Tests
    [Fact]
    public void TestGetCurrentReadingReturnsLatestData()
    {
        _station.RegisterObserver(_observer1);
        _station.SetWeatherData(25, 70, 1015, 15);

        var reading = _observer1.LastStation.GetCurrentReading();
        Assert.Equal(25, reading.Temperature);
        Assert.Equal(70, reading.Humidity);
        Assert.Equal(1015, reading.Pressure);
        Assert.Equal(15, reading.WindSpeed);
    }

    [Fact]
    public void TestGetReadingsReturnsFullHistory()
    {
        _station.RegisterObserver(_observer1);

        _station.SetWeatherData(20, 60, 1010, 10);
        _station.SetWeatherData(25, 70, 1015, 15);
        _station.SetWeatherData(30, 80, 1020, 20);

        var readings = _observer1.LastStation.GetReadings();
        Assert.Equal(4, readings.Count); // initial + 3 updates
        Assert.Equal(22, readings[0].Temperature);
        Assert.Equal(30, readings[3].Temperature);
    }

    [Fact]
    public void TestMaintainsLast10ReadingsMax()
    {
        _station.RegisterObserver(_observer1);

        for (int i = 0; i < 12; i++)
        {
            _station.SetWeatherData(20 + i, 60, 1010, 10);
        }

        var readings = _observer1.LastStation.GetReadings();
        Assert.Equal(10, readings.Count);
    }

    // ── Concrete Observers Tests
    [Fact]
    public void TestTemperatureChartRegistersAndUpdates()
    {
        var chart = new TemperatureChart("Test Temp");
        _station.RegisterObserver(chart);
        _station.SetWeatherData(25, 70, 1015, 15); // Should not throw
    }

    [Fact]
    public void TestHumidityChartRegistersAndUpdates()
    {
        var chart = new HumidityChart("Test Humidity");
        _station.RegisterObserver(chart);
        _station.SetWeatherData(25, 70, 1015, 15); // Should not throw
    }

    [Fact]
    public void TestPressureWindChartRegistersAndUpdates()
    {
        var chart = new PressureWindChart("Test Pressure");
        _station.RegisterObserver(chart);
        _station.SetWeatherData(25, 70, 1015, 15); // Should not throw
    }

    [Fact]
    public void TestAllThreeCanRegisterTogether()
    {
        var temp = new TemperatureChart("Temp");
        var humid = new HumidityChart("Humidity");
        var pressure = new PressureWindChart("Pressure");

        _station.RegisterObserver(temp);
        _station.RegisterObserver(humid);
        _station.RegisterObserver(pressure);

        _station.SetWeatherData(25, 70, 1015, 15); // Should not throw
    }

    // ── Loose Coupling Tests
    [Fact]
    public void TestObserversAreIndependent()
    {
        var obs3 = new MockObserver();
        _station.RegisterObserver(_observer1);
        _station.RegisterObserver(_observer2);
        _station.RegisterObserver(obs3);

        _station.RemoveObserver(_observer2);
        _station.SetWeatherData(25, 70, 1015, 15);

        Assert.Equal(1, _observer1.UpdateCount);
        Assert.Equal(0, _observer2.UpdateCount); // Never notified
        Assert.Equal(1, obs3.UpdateCount);
    }

    [Fact]
    public void TestDynamicSubscription()
    {
        _station.SetWeatherData(20, 60, 1010, 10); // Before registration

        _station.RegisterObserver(_observer1);
        _station.SetWeatherData(25, 70, 1015, 15); // After registration

        Assert.Equal(1, _observer1.UpdateCount); // Only the second one

        _station.RemoveObserver(_observer1);
        _station.SetWeatherData(30, 80, 1020, 20); // After removal

        Assert.Equal(1, _observer1.UpdateCount); // Still only one
    }

    [Fact]
    public void TestStationDoesNotDependOnConcreteObserverTypes()
    {
        var custom = new CustomObserver();
        _station.RegisterObserver(custom);
        _station.SetWeatherData(25, 70, 1015, 15);

        Assert.True(custom.Called);
    }
}


public class CustomObserver : Observer
{
    public bool Called { get; set; } = false;

    public void Update(WeatherStation station)
    {
        Called = true;
    }
}

File: patterns/_tests/BehavioralPatterns/Observer/java/unit.test.java
// ============================================================
//  Unit Tests — Observer: Weather Station (Java)
//
//  What we test here:
//    · Observer registration/removal mechanics
//    · Notification broadcast to all registered observers
//    · Pull model — observers get current readings
//    · Reading history maintained correctly
//    · Observer independence — removal doesn't affect others
//
//  Run: javac -cp .:junit-4.13.2.jar:hamcrest-core-1.3.jar ObserverUnitTest.java && java -cp .:junit-4.13.2.jar:hamcrest-core-1.3.jar org.junit.runner.JUnitCore ObserverUnitTest
// ============================================================

import org.junit.*;
import static org.junit.Assert.*;
import java.util.*;


public class UnitTest {
    private WeatherStation station;
    private MockObserver observer1;
    private MockObserver observer2;

    @Before
    public void setUp() {
        station = new WeatherStation();
        observer1 = new MockObserver();
        observer2 = new MockObserver();
    }

    // ── Observer Registration Tests
    @Test
    public void testRegistersASingleObserver() {
        station.registerObserver(observer1);
        assertEquals(0, observer1.updateCount); // No update yet
    }

    @Test
    public void testPreventsDuplicateRegistration() {
        station.registerObserver(observer1);
        station.registerObserver(observer1); // Try again
        station.setWeatherData(20, 60, 1010, 10);
        assertEquals(1, observer1.updateCount); // Called only once
    }

    @Test
    public void testRemovesAnObserverFromTheList() {
        station.registerObserver(observer1);
        station.registerObserver(observer2);
        station.removeObserver(observer1);
        station.setWeatherData(20, 60, 1010, 10);
        assertEquals(0, observer1.updateCount); // Not notified
        assertEquals(1, observer2.updateCount); // Still notified
    }

    // ── Notification Broadcast Tests
    @Test
    public void testNotifiesAllRegisteredObservers() {
        MockObserver obs3 = new MockObserver();
        station.registerObserver(observer1);
        station.registerObserver(observer2);
        station.registerObserver(obs3);

        station.setWeatherData(25, 70, 1015, 15);

        assertEquals(1, observer1.updateCount);
        assertEquals(1, observer2.updateCount);
        assertEquals(1, obs3.updateCount);
    }

    @Test
    public void testNotifiesRepeatedlyOnMultipleUpdates() {
        station.registerObserver(observer1);

        station.setWeatherData(20, 60, 1010, 10);
        assertEquals(1, observer1.updateCount);

        station.setWeatherData(25, 70, 1015, 15);
        assertEquals(2, observer1.updateCount);

        station.setWeatherData(30, 80, 1020, 20);
        assertEquals(3, observer1.updateCount);
    }

    @Test
    public void testPassesTheStationItselfToUpdate() {
        station.registerObserver(observer1);
        station.setWeatherData(25, 70, 1015, 15);

        assertSame(observer1.lastStation, station);
    }

    // ── Pull Model Tests
    @Test
    public void testGetCurrentReadingReturnsLatestData() {
        station.registerObserver(observer1);
        station.setWeatherData(25, 70, 1015, 15);

        WeatherData reading = observer1.lastStation.getCurrentReading();
        assertEquals(25, reading.temperature, 0.01);
        assertEquals(70, reading.humidity, 0.01);
        assertEquals(1015, reading.pressure, 0.01);
        assertEquals(15, reading.windSpeed, 0.01);
    }

    @Test
    public void testGetReadingsReturnsFullHistory() {
        station.registerObserver(observer1);

        station.setWeatherData(20, 60, 1010, 10);
        station.setWeatherData(25, 70, 1015, 15);
        station.setWeatherData(30, 80, 1020, 20);

        List<WeatherData> readings = observer1.lastStation.getReadings();
        assertEquals(4, readings.size()); // initial + 3 updates
        assertEquals(22, readings.get(0).temperature, 0.01); // initial
        assertEquals(30, readings.get(3).temperature, 0.01); // latest
    }

    @Test
    public void testMaintainsLast10ReadingsMax() {
        station.registerObserver(observer1);

        for (int i = 0; i < 12; i++) {
            station.setWeatherData(20 + i, 60, 1010, 10);
        }

        List<WeatherData> readings = observer1.lastStation.getReadings();
        assertEquals(10, readings.size());
    }

    // ── Concrete Observers Tests
    @Test
    public void testTemperatureChartRegistersAndUpdates() {
        TemperatureChart chart = new TemperatureChart("Test Temp");
        station.registerObserver(chart);
        station.setWeatherData(25, 70, 1015, 15); // Should not throw
    }

    @Test
    public void testHumidityChartRegistersAndUpdates() {
        HumidityChart chart = new HumidityChart("Test Humidity");
        station.registerObserver(chart);
        station.setWeatherData(25, 70, 1015, 15); // Should not throw
    }

    @Test
    public void testPressureWindChartRegistersAndUpdates() {
        PressureWindChart chart = new PressureWindChart("Test Pressure");
        station.registerObserver(chart);
        station.setWeatherData(25, 70, 1015, 15); // Should not throw
    }

    @Test
    public void testAllThreeCanRegisterTogether() {
        TemperatureChart temp = new TemperatureChart("Temp");
        HumidityChart humid = new HumidityChart("Humidity");
        PressureWindChart pressure = new PressureWindChart("Pressure");

        station.registerObserver(temp);
        station.registerObserver(humid);
        station.registerObserver(pressure);

        station.setWeatherData(25, 70, 1015, 15); // Should not throw
    }

    // ── Loose Coupling Tests
    @Test
    public void testObserversAreIndependent() {
        MockObserver obs3 = new MockObserver();
        station.registerObserver(observer1);
        station.registerObserver(observer2);
        station.registerObserver(obs3);

        station.removeObserver(observer2);
        station.setWeatherData(25, 70, 1015, 15);

        assertEquals(1, observer1.updateCount);
        assertEquals(0, observer2.updateCount); // Never notified
        assertEquals(1, obs3.updateCount);
    }

    @Test
    public void testDynamicSubscription() {
        station.setWeatherData(20, 60, 1010, 10); // Before registration

        station.registerObserver(observer1);
        station.setWeatherData(25, 70, 1015, 15); // After registration

        assertEquals(1, observer1.updateCount); // Only the second one

        station.removeObserver(observer1);
        station.setWeatherData(30, 80, 1020, 20); // After removal

        assertEquals(1, observer1.updateCount); // Still only one
    }

    @Test
    public void testStationDoesNotDependOnConcreteObserverTypes() {
        class CustomObserver implements Observer {
            boolean called = false;
            @Override
            public void update(WeatherStation station) {
                called = true;
            }
        }

        CustomObserver custom = new CustomObserver();
        station.registerObserver(custom);
        station.setWeatherData(25, 70, 1015, 15);

        assertTrue(custom.called);
    }
}

// Test Observer (package-private, not a test class)
class MockObserver implements Observer {
    int updateCount = 0;
    WeatherStation lastStation = null;

    @Override
    public void update(WeatherStation station) {
        updateCount++;
        lastStation = station;
    }
}
